File: four/part1.rb
require 'set'

sum=0
while (line = gets)
  winners, numbers = line.chomp.match(/: (.*?) \| (.*)/).captures
  winners = Set.new(winners.split.map(&:to_i))
  sum += numbers.split.map(&:to_i).select { |n| winners.include?(n) }.reduce(0) { |total, n|  total.zero? ? 1 : total * 2 }
end
puts sum

File: four/part1.py
import sys
import re
from functools import reduce

sum=0
for line in sys.stdin.readlines():
    winners,numbers = re.match(r'.*: (.*?) \| (.*)', line).groups()
    winners = set([int(x) for x in winners.split()])
    sum += reduce(lambda total, n: 1 if total == 0 else total*2, filter(lambda x: x in winners,[int(x) for x in numbers.split()]),0)

print(sum)
